content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,488 @@
+```
+
+```console
+pod "pod" deleted
+```
+
+```shell
+kubectl apply -f pod.yaml
+```
+
+```console
+pod/pod created
+```
+
+```console
+INFO[…] Found image specific seccomp profile annotation: seccomp-profile.kubernetes.cri-o.io=quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Pulling OCI artifact from ref: quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Trying to read profile from: /var/lib/containers/storage/overlay/c15f3eb1efd9abe0ae973e050c374988bba52c45a53f5c3cd80419a865013fa6/merged/seccomp.json  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Retrieved OCI artifact seccomp profile of len: 18853  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Created container 116a316cd9a11fe861dd04c43b94f45046d1ff37e2ed05a4e4194fcaab29ee63: default/pod/container  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+```
+
+```shell
+export CONTAINER_ID=$(sudo crictl ps --name container -q)
+sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp | head
+```
+
+```yaml
+{
+  "defaultAction": "SCMP_ACT_ERRNO",
+  "defaultErrnoRet": 38,
+  "architectures": [
+    "SCMP_ARCH_X86_64",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X32"
+  ],
+  "syscalls": [
+    {
+```
+
+For container images, the annotation `seccomp-profile.kubernetes.cri-o.io` will
+be treated in the same way as `seccomp-profile.kubernetes.cri-o.io/POD` and
+applies to the whole pod. In addition to that, the whole feature also works when
+using the container specific annotation on an image, for example if a container
+is named `container1`:
+
+```shell
+skopeo inspect --raw docker://quay.io/crio/nginx-seccomp:container |
+    jq '.annotations."seccomp-profile.kubernetes.cri-o.io/container1"'
+```
+
+```console
+"quay.io/crio/seccomp:v1"
+```
+
+The cool thing about this whole feature is that users can now create seccomp
+profiles for specific container images and store them side by side in the same
+registry. Linking the images to the profiles provides a great flexibility to
+maintain both over the whole application's life cycle.
+
+### Pushing profiles using ORAS
+
+The actual creation of the profiles requires a bit more prework when using ORAS.
+I have the hope that tools like Podman will simplify the overall process in the
+future. Right now, the container registry needs to be
+[ORAS compatible](https://oras.land/docs/compatible_oci_registries/#registries-supporting-oci-artifacts),
+which is also (partially) the case for [Quay.io](https://quay.io). CRI-O also
+needs a `config` with the `mediaType` `application/vnd.oci.image.config.v1+json`,
+while ORAS uses `application/vnd.oci.empty.v1+json` per default. Last but not
+least, CRI-O requires a tarball rather than the plain `.json` file. To achieve
```

Comment:
if I've understood correctly, it's the [OCI image spec](https://github.com/opencontainers/image-spec/blob/main/spec.md) that requires tar, or gzipped tar as the format of a layer (and then CRI-O's technical choice to use container images for seccomp profile distribution implies that the image has to follow the spec for images).

```suggestion
least, the OCI format for container images requires that the profile (in JSON) is encoded as a
tar-fomat container image layer. To achieve
```

[If CRI-O or some other project used a dedicated format for distributing the profiles as OCI objects, there wouldn't be any need for using tar here.]